app/services/proforma_service/convert_task_to_exercise.rb

```diff
@@ -60,44 +68,55 @@ def string_to_bool(str)
     end
 
     def files
-      model_solution_files + test_files + task_files.values.tap {|array| array.each {|file| file.role ||= 'regular_file' } }
+      model_solution_files + test_files + task_files
     end
 
     def test_files
-      @task.tests.map do |test_object|
-        task_files.delete(test_object.files.first.id).tap do |file|
-          file.weight = extract_meta_data(test_object.meta_data&.dig('test-meta-data'), 'weight').presence || 1.0
-          file.feedback_message = extract_meta_data(test_object.meta_data&.dig('test-meta-data'), 'feedback-message').presence || 'Feedback'
-          file.hidden_feedback = extract_meta_data(test_object.meta_data&.dig('test-meta-data'), 'hidden-feedback').presence || false
-          file.role ||= 'teacher_defined_test'
+      @task.tests.flat_map do |test|
+        test.files.map do |task_file|
+          codeocean_file_from_task_file(task_file, test).tap do |file|
+            file.weight = extract_meta_data(test.meta_data&.dig('test-meta-data'), 'test-file', task_file.id, 'weight').presence || 1.0
+            file.feedback_message = extract_meta_data(test.meta_data&.dig('test-meta-data'), 'test-file', task_file.id, 'feedback-message').presence || 'Feedback'
+            file.hidden_feedback = extract_meta_data(test.meta_data&.dig('test-meta-data'), 'test-file', task_file.id, 'hidden-feedback').presence || false
+            file.role = 'teacher_defined_test' unless file.teacher_defined_assessment?
+          end
         end
       end
     end
 
     def model_solution_files
-      @task.model_solutions.map do |model_solution_object|
-        task_files.delete(model_solution_object.files.first.id).tap do |file|
-          file.role ||= 'reference_implementation'
+      @task.model_solutions.flat_map do |model_solution|
+        model_solution.files.map do |task_file|
+          codeocean_file_from_task_file(task_file, model_solution).tap do |file|
+            file.role ||= 'reference_implementation'
+            file.feedback_message = nil
+          end
         end
       end
     end
 
     def task_files
-      @task_files ||= @task.all_files.reject {|file| file.id == 'ms-placeholder-file' }.to_h do |task_file|
-        [task_file.id, codeocean_file_from_task_file(task_file)]
+      @task.files.reject {|file| file.id == 'ms-placeholder-file' }.map do |task_file|
+        codeocean_file_from_task_file(task_file).tap do |file|
+          file.role ||= 'regular_file'
+          file.feedback_message = nil
+        end
       end
     end
 
-    def codeocean_file_from_task_file(file)
+    def codeocean_file_from_task_file(file, parent_object = nil)
       extension = File.extname(file.filename)
-      codeocean_file = CodeOcean::File.new(
+      # checking the last element of xml_id_path array for file.id
+      codeocean_file = @exercise.files.where('array_length(xml_id_path, 1) IS NOT NULL AND xml_id_path[array_length(xml_id_path, 1)] = ?', file.id).first_or_initialize
```

Comment:
This SQL query is executed per file, thus creating an N+1 query. Rather than using SQL here, it would have been more beneficial to use in-memory filtering with Ruby. That way, we would load all files at once (using `@exercise.files`) and then simply select the matching file. I would expect that this improves performance (less SQL) and potentially readability (easier access to last array element). Since we access `@exercise.files` in `destroy_old_files` already, this would even skip the first SQL query (to get all files). 

I see this part as optimization only and, given the infrequent use of the endpoint and the time spent already, would be fine to skip this optimization here.